### PR TITLE
add minProperties and maxProperties to object

### PR DESCRIPTION
**Pull request by amitu**

### CODE REVIEW

**Comment by amitu** [comment]
I have a question: the goal of this project is to make it impossible to have invalid schema. What is I pass 4  required `properties`, and yet say `maxProperties` is 3, currently the schema would be created, but its non-sense. No real JSON can satisfy this schema. 

You think we should try to detect this:

```elm
object : List ObjectSchemaProperty -> Schema
object props =
    -- TODO: figure out how many required properties are, and what is maxProperties,
    -- and ensure its not contradictory.
    -- 
    -- Similarly ensure minProperties < maxProperties
    List.foldl (<|) defaultObject props
        |> Object
```

**Comment by jwoudenberg** [comment]
Hey @amitu, thanks for this! One question: would it be possible to add these fields to the decoder and fuzzer as well?

**Comment by amitu** [comment]
@jwoudenberg:

1. I think I already have added them to decoder. 

2. Have improved fuzzer on booleanSchema.enum.

2. Stared at the definition of `objectFuzzer` till my eyes hurt, but could not wrap my head around what to do. Any hints?

File: tests/specs/SchemaFuzzSpec.elm [hunk 1]
```diff
@@ -31,14 +31,22 @@ spec =
             , array [ items (string []), minItems 5 ]
             , array [ items (string []), maxItems 10 ]
             , array [ items (string []), minItems 5, maxItems 10 ]
+            , object
```
Comment (jwoudenberg) [inline tests/specs/SchemaFuzzSpec.elm]: It's working because the `JsonSchema.Validator` module is not yet checking these rules.

**Comment by jwoudenberg** [comment]
Sorry for the slow response and happy to give some pointers!

The basic idea is that for each property schema the user has defined, we generate 0 or 1 values. We always generate a value if the property is required, and we may or may not generate a value if it's optional.

So after [this line](https://github.com/NoRedInk/json-elm-schema/blob/3fc32c8f5a6b525b7f4d3df89ffcbe313bdf7c0d/src/JsonSchema/Fuzz.elm#L74) we end up with a list of maybe fuzzers (one for each property): `List (Fuzzer (Maybe a))`.

The [`Fuzz.Extra.sequence`](https://github.com/NoRedInk/json-elm-schema/blob/3fc32c8f5a6b525b7f4d3df89ffcbe313bdf7c0d/src/JsonSchema/Fuzz.elm#L75) inverts the `List` and `Fuzzer` in the type above, so we get a fuzzer of lists: `Fuzzer (List (Maybe a))`.

In the [last line](https://github.com/NoRedInk/json-elm-schema/blob/3fc32c8f5a6b525b7f4d3df89ffcbe313bdf7c0d/src/JsonSchema/Fuzz.elm#L76) we reach inside the `Fuzzer` and first call `Maybe.Extra.values` to remove all `Nothing` values and only keep the `Just`'s, then we have our list of properties and encode the result.

Looking at it now, adding `minProperties` and `maxProperties` to this looks like it would be an interesting challenge:
- Instead of just letting the `maybe` fuzzer decide whether we to include an optional schema, we'll need to take the `minProperties` and `maxProperties` into account to ensure we end up with enough properties.
- We have to worry about the possibility that the task we set ourselves is impossible, which you brought up in another issue.
- If the user hasn't set `additionalProperties` to false, that's basically a carte-blanch for us to start making up random extra values. This is something we're currently not doing, but we might need to start doing in order to generate enough enough properties to match a `minProperties` requirement.

It would be nice to be able to merge without all that, so I propose we go the simpler way for now and just not support `minProperties` and `maxProperties` in the Fuzzer case. We're currently using `Debug.Crash` when we detect a schema is using properties unsupported by Fuzzing, so it makes sense to do that here too if we notice `minProperties` or `maxProperties` are present. (It would be nice to switch to using [`Fuzz#invalid`](http://package.elm-lang.org/packages/elm-community/elm-test/4.1.0/Fuzz#invalid) at some point though).

**Comment by amitu** [comment]
> It would be nice to be able to merge without all that, so I propose we go the simpler way for now and just not support minProperties and maxProperties in the Fuzzer case.

The part in me who has to use it in my work already says "YES please!", the other will not forgive me for giving up on a challenge :-P

Good thing *you* have to take the call.

**Comment by jwoudenberg** [comment]
> The part in me who has to use it in my work already says "YES please!", the other will not forgive me for giving up on a challenge :-P
> 
> Good thing you have to take the call.

The challenge is still available! I'm only saying we can merge without addressing it straight away ;).

**Comment by amitu** [comment]
:-) Lets merge for now then.

**Comment by jwoudenberg** [comment]
Cool!

**Comment by jwoudenberg** [comment]
2.1.0 released!

**Comment by amitu** [comment]
maxProperties-ed all the objects. 

JSON Schema should have some way to specify only these properties can be sent, no other (with max properties, if some optional property is not being sent, the client may be sending something else, e.g. they had a typo in one of the optional keys: we won't be able to catch it).

**Comment by jwoudenberg** [comment]
I might misunderstand, but I think that's what `additionalProperties: false` is for?

**Comment by amitu** [comment]
Not sure.

> The value of "additionalProperties" MUST be a valid JSON Schema.
> 
> This keyword determines how child instances validate for objects, and does not directly validate the immediate instance itself.
>
> Validation with "additionalProperties" applies only to the child values of instance names that do not match any names in "properties", and do not match any regular expression in "patternProperties".
>
> For all such properties, validation succeeds if the child instance validates against the "additionalProperties" schema.
>
> Omitting this keyword has the same behavior as an empty schema.

Doesn't say it can be `false`. 

But I think it can be used as a hack, say we put additionalProperty as a bool, with enum containing no value.

**Comment by jwoudenberg** [comment]
Wow, that's interesting. It used to be possible to be set to `false` for v4 schema's, but it seems that got changed. It seems you're now recommended to use [`propertyNames`](http://json-schema.org/latest/json-schema-validation.html#rfc.section.6.22) to limit valid properties.

**Comment by jwoudenberg** [comment]
Wait, nope, actually it's still valid, because `true` and `false` have been made valid schema's everywhere (and so pass the definition you quoted above): https://github.com/json-schema-org/json-schema-spec/issues/133

**Pull request merged**